### DISCUSSION TITLE
Err(UnrecognizedScript)

### DISCUSSION THREAD

**Discussion by gd87429** [rust-bitcoin/rust-bitcoin#1398] [Q&A] [answered]
Sorry, I feel like this might be a dumb question, but I was looking at tx `f4184fc596403b9d638783cf57adfe4c75c605f6356fbc91338530e9831e9e16` and noticed I've been getting `Err(UnrecognizedScript)` when I try to use:

```
Address::from_script(script, network);
```

For network I'm using `bitcoin::network::constants::Network::Bitcoin` and script seems to be:

```
Script(OP_PUSHBYTES_65 0411db93e1dcdb8a016b49840f8c53bc1eb68a382e97b1482ecad7b148a6909a5cb2e0eaddfb84ccf9744464f82e160bfa9b8b64f9d4c03f999b8643f656b412a3 OP_CHECKSIG)
```

Is there something different about early transactions? How can I generate that pubic address?

**Comment by sanket1729** [comment]
Those scripts don't have any address format associated with it.

**Reply by gd87429** [reply]
Hm, ok. So how would one go about converting it to an address like `12cbQLTFMXRnSzktFkuoG3eHoMeFtpTu3S`?

Edit: From my (limited) understanding, there's a signature that's visible in this script. And the next transaction that references this utxo, will contain more information in TxIn from which the public hash can be calculated? Is that how it works & if yes, any guidance on how I can use this crate to figure out the actual public address? :sweat_smile:

**Comment by gd87429** [comment]
After [digging around a bit more](https://docs.rs/bitcoin/latest/src/bitcoin/util/address.rs.html#1056-1066), something like this will produce a public address:

```
let key = hex_key!("0411db93e1dcdb8a016b49840f8c53bc1eb68a382e97b1482ecad7b148a6909a5cb2e0eaddfb84ccf9744464f82e160bfa9b8b64f9d4c03f999b8643f656b412a3");
Address::p2pkh(&key, bitcoin::Network::Bitcoin); // 12cbQLTFMXRnSzktFkuoG3eHoMeFtpTu3S
```

Internally `Address::from_script` uses `script.is_p2pkh()` which returns `false` (which is why I was getting `Error::UnrecognizedScript`).

However, this returns `true`:

```
Address::from_str("12cbQLTFMXRnSzktFkuoG3eHoMeFtpTu3S")?.address_type().unwrap() == AddressType::P2pkh; // true
```

Which makes me conclude that from script, it's not possible to determine address type, but from public address, it is... Somehow seems inconsistent, library-wise..?

> Those scripts don't have any address format associated with it.

Is it dangerous to assume that all early txs are p2pkh?

Basically, I'm trying to figure out a solid way to get the input/output addresses for something as early as block 170, and I'm having a hard time doing that.

**Reply by apoelstra** [reply]
>Is it dangerous to assume that all early txs are p2pkh?

Yes, because they are not p2pkh. You can see in the script that you posted that there is no hashing anywhere.

> Basically, I'm trying to figure out a solid way to get the input/output addresses for something as early as block 170, and I'm having a hard time doing that.

There are no addresses associated with p2pk. The "addresses" that were used at that time are ambiguous with p2pkh and are no longer meaningful, and **will be misinterpreted by any wallet you give them to**.

**Reply by gd87429** [reply]
Got it, thank you for elaborating.

I'm having trouble getting anyone to help me with my original question: how to find a public address from looking at the transaction from block 170. For example, [this block explorer](https://www.blockchain.com/btc/tx/f4184fc596403b9d638783cf57adfe4c75c605f6356fbc91338530e9831e9e16) has figured it out, but I can't seem to figure it out with this library.

Is the answer that it's not possible to do it accurately or maybe this is the wrong library for doing that? I'm fine with that, I could just use some guidance as to whether I should give up on **figuring out the no-longer-meaningful address** or if there is a way to retrieve it.

**Reply by apoelstra** [reply]
The block explorer you linked is lying to you, and has been telling lies since at least as far back 2014. It has not "figured it out".

>Is the answer that it's not possible to do it accurately or maybe this is the wrong library for doing that? I'm fine with that, I could just use some guidance as to whether I should give up on figuring out the no-longer-meaningful address or if there is a way to retrieve it.

It's possible to emulate what blockchain.info is doing, using this library, but I don't understand what you're actually trying to accomplish, because the result will be wrong in a potentially money-losing way.

**Reply by gd87429** [reply]
Ahh, got it got it got it. So my understanding is, they're treating it as if it's one type of address, when in reality it's not, and there is no "proper / accurate" address.

> I don't understand what you're actually trying to accomplish

I was trying to emulate a block explorer, where I can see how funds flow, but since the code was giving me an error when trying to figure out the *to* address (as I now see correctly), I could not trace the flow of coins because the trail ended there (no address to tie it with). So, as you can see, for this use-case I'm more concerned about assigning a unique address, rather than the address being cryptographically accurate.

Anyway, I won't waste any more of your time. Thank you so much for taking the time 🙏

**Comment by apoelstra** [comment, marked as answer]
>Ahh, got it got it got it. So my understanding is, they're treating it as if it's one type of address, when in reality it's not, and there is no "proper / accurate" address.

Bingo! And the concern here is that modern wallets (since 2011 at least..) will interpret the address in the "modern" way, which won't match what the owner of the funds' wallet will do. So you could send coins and they may not be recognized and may not even be spendable [1].

>So, as you can see, for this use-case I'm more concerned about assigning a unique address, rather than the address being cryptographically accurate.

Gotcha. So, my advice would be to invent your own "address" scheme where you output "`pk(<public key in hex>)`". (In fact, you may be able to use `rust-miniscript` to do this for you, though I'm not immediately sure how.) You'll have to manually check for pay-to-pubkey addresses using `Script::is_p2pk`, parse out the key [2]

If you want to match blockchain.info exactly, you can use the hack that you described, where you pull out the public key and then stick it into `Address::p2pkh()`, but the result won't be "unique" and may be dangerous for the above reasons.

[1] In practice they almost certainly would be spendable, with enough hacking, since all the wallets of the era had key material stored on disk which you could access and write new code around. But this is really playing with fire.
[2] We should really have a method on `Script` that does this parsing for you

**Reply by gd87429** [reply]
Wow, great info. Thank you for the suggestions. Makes so much sense now.
